src/main/java/de/tum/in/www1/artemis/repository/StudentParticipationRepository.java

```diff
@@ -1210,4 +1210,33 @@ SELECT COALESCE(AVG(p.presentationScore), 0)
                 AND p.presentationScore IS NOT NULL
             """)
     double getAvgPresentationScoreByCourseId(@Param("courseId") long courseId);
+
+    @Query("""
+            SELECT DISTINCT p
+            FROM StudentParticipation p
+                LEFT JOIN FETCH p.results r
+                LEFT JOIN FETCH r.submission s
+                LEFT JOIN FETCH r.feedbacks f
+                LEFT JOIN FETCH f.testCase
+            WHERE p.exercise.id = :exerciseId
+                AND (
+                    r.id = (SELECT MAX(p_r.id) FROM p.results p_r)
+                    OR r.assessmentType <> de.tum.in.www1.artemis.domain.enumeration.AssessmentType.AUTOMATIC
+                    OR r IS NULL
+                )
+            """)
+    Set<StudentParticipation> findByExerciseIdWithLatestResultsAndFeedbackAndTestcases(@Param("exerciseId") long exerciseId);
+
+    /**
+     * Get the latest results and their feedbacks for a given exercise ID from the database (independant of if the correction was manual or automatic).
+     * The results are loaded together with their feedback and testcases.
+     * Throws an EntityNotFoundException if no results could be found for the given exercise ID.
+     *
+     * @param exerciseId the ID of the exercise for which the latest results and feedbacks should be loaded from the database
+     * @return the set of student participations containing the latest results and their feedbacks
+     */
+    default Set<StudentParticipation> findByExerciseIdWithLatestResultsAndFeedbackAndTestcasesElseThrow(long exerciseId) {
+        return getArbitraryValueElseThrow(Optional.of(findByExerciseIdWithLatestResultsAndFeedbackAndTestcases(exerciseId)), "Results with Feedback for: " + exerciseId);
```

Comment:
Two things:
1. Will `findByExerciseIdWithLatestResultsAndFeedbackAndTestcases` ever return `null`? In case that there are no results, I would expect it to return an empty Set.
2. If you want an Optional of a possibly nullable value, you cannot use `of`: "value - the value to be present, which must be non-null". Instead you should use `ofNullable`.

From my standpoint this method is redundant and you could directly use `findByExerciseIdWithLatestResultsAndFeedbackAndTestcases`.